### ISSUE TITLE
How to do multi-task training with multiple adapters & heads? 

### ISSUE THREAD

**Issue by JH-lee95** [adapter-hub/adapter-transformers#461]
Hi I'm trying to solve a multi-task problem (4 tasks, so need 4 labels in one forward)

I set up 4 adapter & head pairs and each of the pairs should take its target label in forward.

But I don't know how to provide exactly 1 label out of 4 labels to a pair.

Should I modify forward method in model class?

(Also, I'm using Huggingface trainer.)

code below is my torch dataset.

It provides input_dict including 4 labels 

```
class CatDataset(Dataset):
    def __init__(self,args,df,tokenizer,pred=False):
        self.tokenizer=tokenizer
        self.df=df
        
        #input sentence
        self.sent=self.df["sent"]
        
        self.max_len=args.max_len
        
        self.pred=pred
        if not self.pred:
            #labels
            self.cat=self.df["cat"]
            self.pol=self.df["pol"]
            self.tense=self.df['tense']
            self.cert=self.df["cert"]
        
    def __len__(self):
        return len(self.df)

    def __getitem__(self,idx):
        
        input_dict=self.tokenizer(self.sent[idx],padding="max_length",max_length=self.max_len,truncation=True,return_tensors="pt")
        
        
        if not self.pred:
            #input_dict["cat"]=torch.tensor(self.cat[idx])
            #input_dict["pol"]=torch.tensor(self.pol[idx])
            #input_dict["tense"]=torch.tensor(self.tense[idx])
            #input_dict["cert"]=torch.tensor(self.cert[idx])
            
            input_dict["labels"]=[torch.tensor(self.cat[idx]),
                                  torch.tensor(self.pol[idx]),
                                  torch.tensor(self.tense[idx]),
                                  torch.tensor(self.cert[idx])
                                  ]
        
        return input_dict
```

**Comment by hSterz** [comment]
Hey, unfortunately, we don't cover that use case. So you need to adjust the forward pass and maybe the composition block. If you want to train all 4 adapters in one forward pass there are two things to consider:
1. Each head needs the correct label to compute the loss. You probably have to adjust the logic [here](https://github.com/adapter-hub/adapter-transformers/blob/master/src/transformers/adapters/heads/base.py#L787) that selects the labels for each head.
2. Keep in mind that the `Parallel` composition block is only supported for inference and not for training ([docs](https://docs.adapterhub.ml/adapter_composition.html#parallel)).
I hope this helps.

**Comment by JH-lee95** [comment]
> Hey, unfortunately, we don't cover that use case. So you need to adjust the forward pass and maybe the composition block. If you want to train all 4 adapters in one forward pass there are two things to consider:
> 
> 1. Each head needs the correct label to compute the loss. You probably have to adjust the logic [here](https://github.com/adapter-hub/adapter-transformers/blob/master/src/transformers/adapters/heads/base.py#L787) that selects the labels for each head.
> 2. Keep in mind that the `Parallel` composition block is only supported for inference and not for training ([docs](https://docs.adapterhub.ml/adapter_composition.html#parallel)).
>    I hope this helps.

Thank you for your suggestion!

It seems I should adjust the forward method

**Comment by adapter-hub-bert** [comment]
This issue has been automatically marked as stale because it has been without activity for 90 days. This issue will be closed in 14 days unless you comment or remove the stale label.

**Comment by adapter-hub-bert** [comment]
This issue was closed because it was stale for 14 days without any activity.
